Added "Network/Genre" View Option to folders

candry7731 opened the pull request:
Added "Network" to the list of view options. Now you can select "network" in the options "view" menu in any tv show folder and see a list all studios. When clicking a studio the grid will filter all items produced by the selected studio

**Changes**
added "Network" View option
Added NetworkTask file to get all studios

**Issues**
Need to update the network grid to display the logo as a wide poster.... maybe limit the columns to 4...

#5

candry7731 commented:
Clearly me a github do not get along.....

neilsb commented:
Functionally seems to work well.

It looks like there is code for loading Genre's, but I don't see that anywhere in the UI.  Is it supposed to be there?

The logos for the Network aren't being displayed when available.  Is this something we should add?

candry7731 commented:
> Functionally seems to work well.
> 
> It looks like there is code for loading Genre's, but I don't see that anywhere in the UI. Is it supposed to be there? You don't see the option for genre's in the view options? For movie's and tv shows you should see networks and genres in the "view" option
> 
> The logos for the Network aren't being displayed when available. Is this something we should add? I do see logo's on my server but they are "scaletozoom" I think they should be scale to fit however is poster are missing you end up with 2 different poster sizes
> 
> Edit: Just one thing I noticed. If you switch to Network, then back to TV Shows, the "Folder" icon is still displayed at the top right of the TV Show posters. I'll look into this

candry7731 reviewed (commented):
remove debug statement

cewert commented on components/ItemGrid/ItemGrid.brs:
```diff
@@ -39,20 +39,25 @@ sub init()
     m.favorite = "Favorite"
 
     m.loadItemsTask = createObject("roSGNode", "LoadItemsTask2")
+    m.LoadNetworksTask = createObject("roSGNode", "LoadNetworksTask")
+    m.LoadGenreTask = createObject("roSGNode", "LoadGenreTask")
```
I know our code standards aren't the best, and maybe they've changed since I've been away, but variables have always been lowerCamelCase. Just like the task above these.

```suggestion
    m.loadNetworksTask = createObject("roSGNode", "LoadNetworksTask")
    m.loadGenreTask = createObject("roSGNode", "LoadGenreTask")
```

cewert commented on components/ItemGrid/ItemGrid.xml:
```diff
@@ -34,14 +34,17 @@
     <field id="selectedItem" type="node" alwaysNotify="true" />
     <field id="quickPlayNode" type="node" alwaysNotify="true" />
     <field id="imageDisplayMode" type="string" value="scaleToZoom" />
+    <field id="StudioList" type="assocarray" />
```
Same with node fields. These should be lowerCamelCase

neilsb replied:
This field doesn't actually seem to be used anywhere, and I think it can be removed.

```suggestion
```

cewert commented on components/ItemGrid/ItemGridOptions.brs:
```diff
@@ -49,6 +49,8 @@ sub optionsSet()
         end for
         m.menus[0].content = viewContent
         m.menus[0].checkedItem = selectedViewIndex
+
+
```
```suggestion
```

neilsb commented:
Funny thing happens when you go into a library showing the original items (Movies / TV Shows) and then switch to Genre.

 I get all the Genres at top of the Grid, but when I scroll down past them it starts Movies and Collections underneath, but missing out the 1st lot of them.   It like if `x` genres are shown, then the next position in the grid (after the last collection) is movie `x + 1` when viewing Movies.  (i.e.  The 1st lot of movies in the grid are replaced by the genres, then the rest of the movies are shown.

Same happens with TV shows.

It looks like the total item count is not getting updated or something, as the title switches to "Movies (Genres) (1 of 500)".  If I back out and go into the library again with it showing Genres immediately, it works as expected.

neilsb commented on components/ItemGrid/ItemGrid.brs:
```diff
@@ -70,10 +75,11 @@ sub loadInitialItems()
         m.filter = get_user_setting("display.livetv.filter")
     else
-        m.view = invalid
+        'm.view = invalid
```
```suggestion
```

Just delete lines rather than commenting them out.  History is available in source control.

neilsb commented on components/ItemGrid/ItemGrid.brs:
```diff
@@ -167,8 +247,13 @@ sub SetUpOptions()
-        options.views = [{ "Title": tr("Shows"), "Name": "shows" }]
+    else if m.top.parentItem.collectionType = "tvshows" or m.top.parentItem.collectionType = invalid
+        options.views = [
+            { "Title": tr("Shows"), "Name": "shows" },
+            { "Title": tr("Networks"), "Name": "Networks" },
```
Need to add `Shows` and `Networks` to the locale file in `locale\en_US\translations.ts`

candry7731 replied:
done

neilsb commented on components/ItemGrid/ItemGrid.brs:
```diff
@@ -167,8 +247,13 @@ sub SetUpOptions()
+        options.views = [
+            { "Title": tr("Shows"), "Name": "shows" },
+            { "Title": tr("Networks"), "Name": "Networks" },
+            { "Title": tr("Genre"), "Name": "Genre" }
```
```suggestion
            { "Title": tr("Genres"), "Name": "Genres" }
```

Can we change this to `Genres` since the other 2 items in the list are plural.  (Genres is used elsewhere and already in the translation file)

neilsb commented on components/ItemGrid/ItemGrid.brs:
```diff
@@ -551,6 +672,12 @@ sub updateTitle()
     end if
 
+    if m.options.view = "Networks" or m.view = "Networks"
+        m.top.overhangTitle = m.top.parentItem.title + tr(" (Networks)")
```
```suggestion
        m.top.overhangTitle =  "%s (%s)".Format(m.top.parentItem.title, tr("Networks"))
```
The "tr" function performs a full string match to find a translation, so if as add `Networks` to the translation file, this wouldn't match ` (Networks)`.   Better to split it out and only pass `Networks` to be translated.

And with so many parts to the string, makes it cleaner to use the format function.

neilsb commented on components/ItemGrid/ItemGrid.brs:
```diff
@@ -551,6 +672,12 @@ sub updateTitle()
+        m.top.overhangTitle = m.top.parentItem.title + tr(" (Networks)")
+    end if
+    if m.options.view = "Genre" or m.view = "Genre"
+        m.top.overhangTitle = m.top.parentItem.title + tr(" (Genre)")
```
```suggestion
        m.top.overhangTitle =  "%s (%s)".Format(m.top.parentItem.title, tr("Genres"))
```

Same as above

neilsb reviewed (changes requested):
Functionally seems to be working well now apart from one one small issue I see. 

1.  If you go into a library and its showing "Networks", and then you switch to "Genres" then the images in the tiles are not the correct size (looks to be scale to fit instead of scale to zoom).  Doesn't happen when switching to/from shows to genres.

A few code suggestions added to the review, and also a few thoughts here too:

1. The file `spinner-old.png` can just be deleted since it's not used anywhere.  It can be retrieved from source control if it's ever required in the future
2. `LoadItemsTask2`, `LoadNetworksTask` and `LoadGenreTask` are all incredibly similar.  Can't the original task just be updated with the logic added to the other 2 and used, rather than duplicating so much code?
3. In [ItemGrid.brs line 41](https://github.com/candry7731/jellyfin-roku/blob/8b27ffa76239a682eb71b84904339b031258b02e/components/ItemGrid/ItemGrid.brs#L41) rather than creating the 3 tasks here and potentially just using one, could this be moved to if around [465](https://github.com/candry7731/jellyfin-roku/blob/8b27ffa76239a682eb71b84904339b031258b02e/components/ItemGrid/ItemGrid.brs#L465), only creating the task where it's needed and not already created.    If it was just stored in a single variable (i.e. `m.LoadItemsTask`), things like the [if at line 414](https://github.com/candry7731/jellyfin-roku/blob/8b27ffa76239a682eb71b84904339b031258b02e/components/ItemGrid/ItemGrid.brs#L414) could be removed

candry7731 commented:
Reworked and deleted genre and networks tasks. Removed all duplicate code. Thanks @neilsb for the suggestions! This should be much better and manageable moving forward. This will also work well with playlist....

neilsb commented:
Looks like an `end if` is missing causing a build failure

candry7731 commented:
> Looks like an `end if` is missing causing a build failure
Fixed this

neilsb commented:
When going into the Options menu within a library, a number of the options are not being shown based on the translation file  (i.e.  Now showing `VIEW`, `SORT` and `FILTER` whereas it was not all uppercase before.  Likewise the sort Options are now all UpperCase except `Run Time`).

candry7731 commented:
How do we want them? Uppercase or LowerCase?

neilsb commented:
> How do we want them? Uppercase or LowerCase?

Mixed case should be fine.  There are already translations in the locale files for these , it's just that the underscores have been replaced by spaces.

```
    <message>
        <source>DATE_ADDED</source>
        <translation>Date Added</translation>
    </message>
```

> This is how the options were written and it displayed like this on the roku, tr("DATE_ADDED") So I removed the underscore and kept the UpperCase. This was working fine when I tested it.

The translations were not working with the underscores when you tried?  Can you double check that and let me know if it's still failing.  When I change the code back to `tr("DATE_ADDED")` it works fine and outputs `Date Added`


And I'm not seeing the **View** options that were there in previous versions of the PR.

![image](https://user-images.githubusercontent.com/5013482/177096280-59bad797-889a-4d87-a9c9-0dfac8d7132f.png)

candry7731 commented:
All of the sort options display with the underscore in the name. It does not look like the translation is working.... not sure why. 
I have readded the network and genre options. Tested this with movies and series. Also tested the alpha picker with each option. 

Please let me know how to fix the sort option names.

neilsb commented on components/ItemGrid/ItemGridOptions.brs:
```diff
@@ -1,7 +1,7 @@
 
     m.buttons = m.top.findNode("buttons")
-    m.buttons.buttons = [tr("TAB_VIEW"), tr("TAB_SORT"), tr("TAB_FILTER")]
+    m.buttons.buttons = [tr("VIEW"), tr("SORT"), tr("FILTER")]
```
```suggestion
    m.buttons.buttons = [tr("TAB_VIEW"), tr("TAB_SORT"), tr("TAB_FILTER")]
```

Just reverting this line should fix the Tab Names.  The "TranslationKey" for these is set to `TAB_VIEW` which it why it wasn't matching `VIEW,` etc

neilsb commented on components/ItemGrid/ItemGrid.brs:
```diff
@@ -397,8 +443,11 @@ sub onItemFocused()
     ' Load more data if focus is within last 3 rows, and there are more items to load
-    if focusedRow >= m.loadedRows - 5 and m.loadeditems < m.loadItemsTask.totalRecordCount
-        loadMoreData()
+    if focusedRow > 1
```
Why is this now required?

candry7731 replied:
This is deleted in the items count pr #655

neilsb commented:
So 2 comments added.

If we can also revert the changes to `package-lock.json`  I think we're good to merge.

candry7731 commented:
This is deleted in the item counts pr #655

On Thu, Jul 7, 2022 at 5:28 AM Neil Burrows ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In components/ItemGrid/ItemGrid.brs
> <https://github.com/jellyfin/jellyfin-roku/pull/596#discussion_r915717222>
> :
>
> > @@ -397,8 +443,11 @@ sub onItemFocused()
>      SetBackground(m.itemGrid.content.getChild(m.itemGrid.itemFocused).backdropUrl)
>
>      ' Load more data if focus is within last 3 rows, and there are more items to load
> -    if focusedRow >= m.loadedRows - 5 and m.loadeditems < m.loadItemsTask.totalRecordCount
> -        loadMoreData()
> +    if focusedRow > 1
>
> Why is this now required?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jellyfin/jellyfin-roku/pull/596#pullrequestreview-1031347674>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT6PDRE3QXJAO7AD6YFOZ23VS2WMLANCNFSM5V465POQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>


-- 
Chris Andry

462 Browning Loop
Mandeville LA 70448
504-617-2924

candry7731 commented:
> So 2 comments added.
> 
> If we can also revert the changes to `package-lock.json` I think we're good to merge.

I didn't change anything in this file manually..... How to I revert or can you?

neilsb reviewed (approved):
👍

candry7731 commented:
> > How do we want them? Uppercase or LowerCase?
> 
> Mixed case should be fine. There are already translations in the locale files for these , it's just that the underscores have been replaced by spaces.
> 
> ```
>     <message>
>         <source>DATE_ADDED</source>
>         <translation>Date Added</translation>
>     </message>
> ```
> 
> > This is how the options were written and it displayed like this on the roku, tr("DATE_ADDED") So I removed the underscore and kept the UpperCase. This was working fine when I tested it.
> 
> The translations were not working with the underscores when you tried? Can you double check that and let me know if it's still failing. When I change the code back to `tr("DATE_ADDED")` it works fine and outputs `Date Added`
> 
> And I'm not seeing the **View** options that were there in previous versions of the PR.
> 
> ![image](https://user-images.githubusercontent.com/5013482/177096280-59bad797-889a-4d87-a9c9-0dfac8d7132f.png)

This should be fixed as the current code has the correct <source> and <translation>

![image](https://user-images.githubusercontent.com/83685828/180065023-67ee2685-557f-4aad-9d91-e01486af47bc.png)

![image](https://user-images.githubusercontent.com/83685828/180065138-19cb9615-b27d-46f8-ab34-ffe35f0f55da.png)
